### PR TITLE
Add style guide docs and refactor documentation

### PR DESCRIPTION
**Pull request by sven1103**
## Motivation
Nf-core's tool application has grown a lot over the past couple of months, so I thought it would be great to go over the source code and do some polishing and refactoring the docstrings.

To motivate others to contribute to the source code, it is a good idea to have good documentation of the code, which can be easily done with `Sphinx` and `autodoc`. I chose [Google's Python Style Guide](https://github.com/google/styleguide/blob/gh-pages/pyguide.md) for the docstrings, as it is very concise, clean and build's upon PEP.  

Docstrings best practices in Python are defined in [https://www.python.org/dev/peps/pep-0257/](https://www.python.org/dev/peps/pep-0257/), however this is to basic, and it would be nice to make use of documentation generators that create verbose doc pages in HTML.

Files to update:
- [x] bump_version.py
- [x] create.py
- [x] download.py
- [x] licenses.py
- [x] lint.py
- [x] list.py
- [x] utils.py

## PR checklist
 - [x] This comment contains a description of changes (with reason)
 - [ ] If you've fixed a bug or added code that should be tested, add tests!
 - [x] Documentation in `docs` is updated
 - [x] `CHANGELOG.md` is updated
 - [ ] `README.md` is updated
 - [ ] `CONTRIBUTION.md` is updated

**Learn more about contributing:** https://github.com/nf-core/tools/tree/master/.github/CONTRIBUTING.md

### CODE REVIEW

File: docs/api/conf.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    'sphinx.ext.napoleon']
+
+# Add any paths that contain templates here, relative to this directory.
+templates_path = ['_templates']
+
+# The suffix(es) of source filenames.
+# You can specify multiple suffix as a list of string:
+#
+# source_suffix = ['.rst', '.md']
+source_suffix = '.rst'
```
Comment (ewels) [inline docs/api/conf.py]: Can we used markdown instead please? We use it everywhere else so better to keep consistency if possible..

Comment (sven1103) [reply docs/api/conf.py]: For the index file sure, this is the entry point of the documentation and will not be touched by the automatic creation of documents. For the auto-generated API doc, this is `.rst`. But as this is the part not touched by us anyway (as it is generated automatically from the Python docstrings), I do not see a problem. The final output will be HTML anyway, so I think we can overlook the intermediary `rst` file from the autodocs?

Comment (sven1103) [reply docs/api/conf.py]: What do you think?

File: nf_core/lint.py [hunk 2]
```diff
@@ -68,19 +68,41 @@ def run_linting(pipeline_dir, release_mode=False):
...
         )
 
     return lint_obj
 
 
 class PipelineLint(object):
-    """ Object to hold linting info and results """
-
-    def __init__(self, pipeline_dir):
+    """Object to hold linting information and results.
```
Comment (sven1103) [inline nf_core/lint.py]: @ewels this is an example (not complete) docstring in Python style format. The other docstring format convenvtion I found is from NumPy.

File: nf_core/lint.py [hunk 3]
```diff
@@ -68,19 +68,41 @@ def run_linting(pipeline_dir, release_mode=False):
...
+                             the following structure, with some example values::
+
+                                 {'name': 'nf-core-hlatyping',
+                                  'channels': ['bioconda', 'conda-forge'],
+                                  'dependencies': ['optitype=1.3.2', 'yara=0.9.6']}
+
+        conda_package_info (dict):
+        config (dict):
+        dockerfile (list):
+        failed (list):
```
Comment (sven1103) [inline nf_core/lint.py]: of course if there is some convention based on Markdown, from which API docs can be generated, I am happy to give it a try.

File: nf_core/lint.py (hunk 2 above)
Comment (sven1103) [reply nf_core/lint.py]: of course if there is some convention based on Markdown, from which API docs can be generated, I am happy to give it a try.

**Comment by sven1103** [comment]
Found a package on PyPi [https://pypi.org/project/pydoc-markdown/](https://pypi.org/project/pydoc-markdown/), that enables Python doctoring to markdown conversion. I am really astonished, that there is so little support for docstring conversion.

File: docs/api/conf.py (hunk 1 above)
Comment (ewels) [reply docs/api/conf.py]: Ah ok cool - if it's the auto-generated doc and never manually edited then I guess it doesn't matter..

**Comment by sven1103** [comment]
So to sum this up: there is little Markdown support, and the doc generators I found, are not sustainable solutions.The two biggest docstring markup standards that are out there for Python derived from the NumPy (https://www.numpy.org/devdocs/docs/howto_document.html) project and Google (https://github.com/google/styleguide/blob/gh-pages/pyguide.md).

So I guess we need to make a decision. I think both style guides are very straight forward, plus, that we get a lot of tools for conversion to HTML. In my opinion, that the intermediate files are `*.rst` does not bother me. Another plus is, that we get cross-linking support, meaning that objects, class, methods and stuff get interlinked with href during the conversion, which is powerful for navigation in the final docs.

@ewels @apeltzer any thoughts?

**Comment by apeltzer** [comment]
I'm happy with both `rst` and `md` files to be honest. I also think, if `rst` is just intermediate, this doesn't matter at all. So I'd say go ahead with `rst` syntax for the documentation....

**Comment by ewels** [comment]
Sure, if it is only intermediate. But will we not be writing it in the docstrings for functions?

**Comment by sven1103** [comment]
@ewels the docstrings syntax from Google or NumPy are own style formats. See here for a comparison:

https://bwanamarko.alwaysdata.net/napoleon/format_exception.html

so far, I haven't seen better docstring syntax style integrations for easy and beautiful api-doc generation. :(

**Comment by sven1103** [comment]
And a docstring syntax based on pure Markdown seems not to have a good support out in the community.

**Comment by sven1103** [comment]
imho, both are not difficult to use. But that is not my call :)

**Comment by ewels** [comment]
This pull request **introduces 1 alert** when merging 83a4b5dbb3c6e4ebca1e83695712b4fe5d0e0c1f into 5bfa9098d28dc3fa72cb11a11a2721df84146185 - [view on LGTM.com](https://lgtm.com/projects/g/nf-core/tools/rev/pr-e8fd43a5c4602f3ca3933a942d6345bdd286e701)

**new alerts:**

* 1 for Wrong number of arguments in a class instantiation

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by ewels** [comment]
This pull request **introduces 1 alert** when merging e2267535aaa4df9ff7ebadd6c73281cf1913687b into 5bfa9098d28dc3fa72cb11a11a2721df84146185 - [view on LGTM.com](https://lgtm.com/projects/g/nf-core/tools/rev/pr-6426f3623db3e01e04783f74ce8d38a25694a6c9)

**new alerts:**

* 1 for Encoding error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by sven1103** [comment]
Ok, the push to the orphan branch does not work yet, as the bot token cannot be resolved in the PR. this will work after merge.

**Comment by sven1103** [comment]
"works on my machine"

**Comment by sven1103** [comment]
Thank you!

**Pull request merged**